File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/HPIndicator.vb
Imports Sakuno.ING.Game.Models

Namespace Controls
    <TemplatePart(Name:="PART_Main", Type:=GetType(RangeBase))>
    <TemplatePart(Name:="PART_Addition", Type:=GetType(RangeBase))>
    Public NotInheritable Class HPIndicator
        Inherits Control

        Public Sub New()
            DefaultStyleKey = GetType(HPIndicator)
        End Sub

        Private _current, _max As Integer
        Public Property HP As ShipHP
            Get
                Return (_current, _max)
            End Get
            Set(value As ShipHP)
                If _max <> value.Max Then
                    _max = value.Max
                    UpdateMax()
                End If
                If _current <> value.Current Then
                    _current = value.Current
                    UpdateCurrent()
                End If
            End Set
        End Property

        Private _additional As Integer?
        Public Property Additional As Integer?
            Get
                Return _additional
            End Get
            Set(value As Integer?)
                If _additional <> value Then
                    _additional = value
                    UpdateAdditional()
                End If
            End Set
        End Property

        Private _additionalBrush As Brush
        Public Property AdditionalBrush As Brush
            Get
                Return _additionalBrush
            End Get
            Set(value As Brush)
                _additionalBrush = value
                If PART_Addition IsNot Nothing Then PART_Addition.Foreground = value
            End Set
        End Property

        Private PART_Main, PART_Addition As RangeBase

        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            PART_Main = TryCast(GetTemplateChild(NameOf(PART_Main)), RangeBase)
            PART_Addition = TryCast(GetTemplateChild(NameOf(PART_Addition)), RangeBase)
            If PART_Main IsNot Nothing Then PART_Main.Minimum = 0
            If PART_Addition IsNot Nothing Then
                PART_Addition.Minimum = 0
                PART_Addition.Foreground = AdditionalBrush
            End If
            UpdateMax()
            UpdateCurrent()
            UpdateAdditional()
        End Sub

        Private Sub UpdateMax()
            If PART_Main IsNot Nothing Then PART_Main.Maximum = _max
            If PART_Addition IsNot Nothing Then PART_Addition.Maximum = _max
        End Sub

        Private Sub UpdateCurrent()
            If PART_Main IsNot Nothing Then PART_Main.Value = _current
            VisualStateManager.GoToState(Me, HP.DamageState.ToString(), True)
        End Sub

        Private Sub UpdateAdditional()
            If PART_Addition IsNot Nothing Then
                PART_Addition.Value = If(_additional, 0)
                PART_Addition.Visibility = If(_additional Is Nothing, Visibility.Collapsed, Visibility.Visible)
            End If
        End Sub
    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/HPIndicator.cs
using Sakuno.ING.Game.Models;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Controls.Primitives;
using Windows.UI.Xaml.Media;

namespace Sakuno.ING.Views.UWP.Controls
{
    [TemplatePart(Name = nameof(PART_Main), Type = typeof(RangeBase))]
    [TemplatePart(Name = nameof(PART_Addition), Type = typeof(RangeBase))]
    public class HPIndicator : Control
    {
        public HPIndicator()
        {
            DefaultStyleKey = typeof(HPIndicator);
        }

        private int _current, _max;
        public ShipHP HP
        {
            get => (_current, _max);
            set
            {
                if (_max != value.Max)
                {
                    _max = value.Max;
                    UpdateMax();
                }
                if (_current != value.Current)
                {
                    _current = value.Current;
                    UpdateCurrent();
                }
            }
        }

        private int? _additional;
        public int? Additional
        {
            get => _additional;
            set
            {
                if (_additional != value)
                {
                    _additional = value;
                    UpdateAdditional();
                }
            }
        }

        private Brush _additionalBrush;
        public Brush AdditionalBrush
        {
            get => _additionalBrush;
            set
            {
                _additionalBrush = value;
                if (PART_Addition != null)
                    PART_Addition.Foreground = value;
            }
        }

        private RangeBase PART_Main, PART_Addition;
        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            PART_Main = GetTemplateChild(nameof(PART_Main)) as RangeBase;
            PART_Addition = GetTemplateChild(nameof(PART_Addition)) as RangeBase;
            if (PART_Main != null) PART_Main.Minimum = 0;
            if (PART_Addition != null)
            {
                PART_Addition.Minimum = 0;
                PART_Addition.Foreground = AdditionalBrush;
            }
            UpdateMax();
            UpdateCurrent();
            UpdateAdditional();
        }

        private void UpdateMax()
        {
            if (PART_Main != null) PART_Main.Maximum = _max;
            if (PART_Addition != null) PART_Addition.Maximum = _max;
        }

        private void UpdateCurrent()
        {
            if (PART_Main != null) PART_Main.Value = _current;
            VisualStateManager.GoToState(this, HP.DamageState.ToString(), true);
        }

        private void UpdateAdditional()
        {
            if (PART_Addition != null)
            {
                PART_Addition.Value = _additional ?? 0;
                PART_Addition.Visibility = _additional == null ? Visibility.Collapsed : Visibility.Visible;
            }
        }
    }
}
